File: questions/q409_maximum_sum_subarray_with_negative_numbers/code_divide_and_conquer.py
class Solution:

    def getMaxSum(self, nums: List[int], s, e) -> List[int] :
        if (s > e) :
            return [0, 0, 0]
        elif s == e :
            self.maxSum = max(self.maxSum, nums[s])
            return [nums[s], nums[s], nums[s]]
        else :
            mid = s + ((e - s)//2)
            left = self.getMaxSum(nums, s, mid)
            right = self.getMaxSum(nums, mid+1, e)

            self.maxSum = max(self.maxSum, left[2]+right[0], left[2], right[0])

            cur = [
                max(left[0], left[1]+right[0]),
                left[1]+right[1],
                max(left[2]+right[1], right[2])
            ]
            return cur

    def maxSubArray(self, nums: List[int]) -> int:
        self.maxSum = float('-inf')
        self.getMaxSum(nums, 0, len(nums)-1)
        return self.maxSum

File: questions/q409_maximum_sum_subarray_with_negative_numbers/code_divide_and_conquer.cpp
class Solution {
public:
    int maxSum;

    Solution() {
        maxSum = INT_MIN;
    }

    vector<int> getMaxSum(vector<int>& nums, int s, int e) {
        if (s > e) {
            return {0, 0, 0};
        }
        else if (s == e) {
            maxSum = max(maxSum, nums[s]);
            return {nums[s], nums[s], nums[s]};
        }
        else {
            int mid = s + ((e - s) / 2);
            vector<int> left = getMaxSum(nums, s, mid);
            vector<int> right = getMaxSum(nums, mid+1, e);

            maxSum = max(maxSum, left[2]+right[0]);
            maxSum = max(maxSum, left[2]);
            maxSum = max(maxSum, right[0]);

            vector<int> cur = {0, 0, 0};
            cur[0] = max(left[0], left[1]+right[0]);
            cur[1] = left[1] + right[1];
            cur[2] = max(left[2]+right[1], right[2]);
            return cur;
        }
    }

    int maxSubArray(vector<int>& nums) {
        int n = nums.size();

        getMaxSum(nums, 0, n-1);
        return maxSum;
    }
};
